### ISSUE TITLE
Stack overflow for some source code input

### ISSUE THREAD

**Issue by JohnnyMorganz** [Kampfkarren/full-moon#140]
For some inputs (eg. https://github.com/JohnnyMorganz/StyLua/issues/38, or other examples shown below), using full-moon produces a stack overflow. This is mostly noticeable in the debug version of full-moon, and is less likely to occur using the release version, but it is still possible.

Some other examples include:
```lua
coroutine.resume(coroutine.create(function(...)
    local succ, err = pcall(tryAPI)
    if not succ then
        local tryConnection; tryConnection = RunService.Stepped:Connect(function(_, delta, ...)
            accum += delta -- stack overflow at this compound assignment
        end)
    else
        
    end
end))
```
[Roblox Feature Flag - if you replace `accum += delta` with something else like `local x = 1`, it parses fine]

```lua
function API.getTags(class: string, member: string?): Array<string>
    if member then
        for _, v in ipairs(classTable.Members) do
            if v.Name == member then
                local tags: typeof(v.Tags) = v.Tags
                if tags then
                    for i, tag in ipairs(tags) do
                        -- stack overflow
                    end
                end
            end
        end
    else
        
    end
end
```
[Roblox Feature Flag - if we remove the first if statemenet and keep its body (`if member then ... else end`), it parses fine. If we remove all type information, but retain the roblox feature flag, there is a stack overflow. If we remove all type information and disable the roblox feature flag, it parses fine]

**Comment by Kampfkarren** [comment]
To be clear, have you ever gotten this on release optimizations, or is that just theoretical? If so, then my potential fix likely won't help.

**Comment by JohnnyMorganz** [comment]
> To be clear, have you ever gotten this on release optimizations, or is that just theoretical? If so, then my potential fix likely won't help. 

https://github.com/JohnnyMorganz/StyLua/issues/38 was done using a release version of StyLua. StyLua also works recursively though, starting at a block and recursively formatting until we reach a TokenReference - this may be compounding onto it?

**Comment by zhiburt** [comment]
Hi @JohnnyMorganz I've tracked a disscussion for a couple of minutes and seems like you know what the issue is about?
But I did not found any place that the issue was actually described.
Or you are aware it exists but you don't know the cause behind?

**Comment by JohnnyMorganz** [comment]
I'm aware it exists, and I know the current workaround is just to use a release build instead.

@Kampfkarren looked more into the underlying cause, and I think they pinpointed something. I can't remember exactly what it was though

**Comment by Kampfkarren** [comment]
Yeah, this is one I'm going to end up taking on my own at some point. I put a thread on the Rust forums and was told to use something called "stacker", which some official Rust products use internally to combat this kind of thing.

**Comment by JohnnyMorganz** [comment]
This seems to be no longer an issue if you update to the latest stable rust version!! 🚀
